src/ekat/kokkos/ekat_kokkos_utils.hpp

```diff
@@ -64,10 +69,80 @@ void parallel_reduce (const TeamMember& team,
 
    result = local_tmp;
   } else {
+    const ValueType initial(result);
     Kokkos::parallel_reduce(Kokkos::TeamThreadRange(team, begin, end), lambda, result);
+    result += initial;
   }
 }
 
+/*
+ * Computes a reduction over a routine described by 'input' for entries 'scalarize(input)([begin,end))'
+ * The variable 'input' should be an operator s.t. 'input(k)' returns a Pack. The computed reduction is
+ * added to the value 'result'.
+ */
+template <bool Serialize, typename TeamMember, typename InputProvider, typename ValueType>
+static KOKKOS_INLINE_FUNCTION
+void view_reduction (const TeamMember& team,
+                     const int begin,
+                     const int end,
+                     const InputProvider& input,
+                     ValueType& result)
+{
+  using PackType = typename std::remove_reference<decltype(input(0))>::type;
+  constexpr int vector_size = PackType::n;
+
+  // Perform a packed reduction over scalar indices
+  if (true) { // !Serialize) {
+    const bool has_garbage_begin = (begin%vector_size != 0 ? true : false);
+    const bool has_garbage_end   = (end%vector_size != 0 ? true : false);
+    const int pack_loop_begin    = (has_garbage_begin ? begin/vector_size + 1 : begin/vector_size);
+    const int pack_loop_end      = end/vector_size;
+
+    // If first pack has garbage, we will not include it in the below reduction,
+    // so manually add the first pack (only the non-garbage part)
+    if (has_garbage_begin) {
+      const int first_indx = begin%vector_size;
+      Kokkos::single(Kokkos::PerTeam(team),[&] {
+        for (int j=first_indx; j<vector_size; ++j) {
+          result += input(pack_loop_begin-1)[j];
```

Comment:
`input` is potentially an expensive lambda. You should evaluate it exactly once per pack. So eval outside the `j` loop, then index a temporary inside the `j` loop.